bootstrap/kubeadm/internal/cloudinit/node_test.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudinit
+
+import (
+	"fmt"
+	"testing"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/yaml"
+)
+
+func TestNewNode(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   *NodeInput
+		check   func([]byte) error
+		wantErr bool
+	}{
+		{
+			"check for duplicated WriteFiles",
+			&NodeInput{
+				BaseUserData: BaseUserData{
+					AdditionalFiles: []bootstrapv1.File{
+						{
+							Path:        "/etc/foo.conf",
+							Content:     "bar",
+							Owner:       "root",
+							Permissions: "0644",
+						},
+					},
+				},
+			},
+			checkWriteFiles("/etc/foo.conf", "/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
+			false,
+		},
+		{
+			"check for existens of /run/kubeadm/kubeadm-join-config.yaml and /run/cluster-api/placeholder",
+			&NodeInput{},
+			checkWriteFiles("/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
+			false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := NewNode(tt.input)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewNode() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if err := tt.check(got); err != nil {
+				t.Errorf("%v: got = %s", err, got)
+			}
+		})
+	}
+}
+
+func checkWriteFiles(files ...string) func(b []byte) error {
+	return func(b []byte) error {
+		var cloudinitData struct {
```

Comment:
nit: Would it make sense to use the struct we're using when marshaling to parse it here again?

Reply:
There's no such struct thing. https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/cloudinit.go#L76

`generate()` uses go templates.